Rules, Transactor preflight, and track STTx validity with HashRouter (RIPD-977)

ximinez opened the pull request:
Note: These changes are based on @vinniefalco's `cache` branch, which is already pending in PR #1187, so the first *five* commits can safely be ignored. The new code starts with "Add Rules to ReadView:".

There are three interdependent, but separate changes in this PR:
1. Rules. Similar to the Fees object, allows looking up transaction processing change rules (ie. the result of Amendment and Change transactions).
2. `preflight`. Checks the transaction without the overhead of reading or trying to apply to a `Ledger`/`OpenView`. Takes an `OpenView`, but only to get the `Rules` object. If successful, returns a function that can be called to apply the transaction to the `OpenView` without duplicating the checks (or allowing an shenanigans caused by changing the inputs).  Application: the transaction queue will use `preflight` to determine whether the transaction is viable _before_ storing it in the queue and broadcasting it to peers.
3. Track `STTx` validity with `HashRouter`. https://ripplelabs.atlassian.net/browse/RIPD-977. Gets rid of the mutable `STTx::sig_state_` member and brings all signature / validity checking into a single source. Also adds a `sigVerify()` function member to the `HashRouter` that can be passed to `preflight` and `apply` to hide the result caching, and prevent duplicate implementations.

ximinez commented:
Reviewers: @nbougalis, @scottschurr

vinniefalco commented:
Notes:

1. This has to go in after the fix to fees in rc4
2. The commit titled "Add Rules to ReadView:" has to be amended, to make sure that the rules are updated in the call to `setImmutable`. This could be as easy as simply fixing the merge conflicts with rc4.

vinniefalco commented:
When this is rebased against *0.29.1* please remove the "always cache on read" commit.

nbougalis commented:
I need justification for why `preflight` returns an invoke lambda. I would prefer folding this commit in https://github.com/nbougalis/rippled/commit/6c4817dbbb186a0d0cc62b669a5338229a90b1f6 which shifts responsibility for the invocation the `apply` where it belongs and simplifies the code a whole lot.

ximinez commented:
Several updates pushed.

@nbougalis Justification for `preflight` returning the lambda:
```
std::pair<TER, bool> TxQ::apply(OpenView& view, shared_ptr<STTx> tx, ...)
{
  [...]
  auto preresult = preflight(tx, ...);
  if( preresult.first != tesSUCCESS )
    return { preresult.first, false };
  auto baseFee = [...]
  auto fee = getFeeLevelPaid(*tx, ...);
  auto requiredFee = loadFeeTrack_.escalateFeeLevel(baseFee);
  if( fee >= requiredFee )
  {
    // Add transaction to the open ledger, using preflight's lambda.
    return preresult.second();
  }
  else 
  {
     // Add transaction to queue, along with any necessary metadata
    [....]
    return { terQUEUED, false };
  }
}

ximinez commented:
Rebased against develop, and squashed. There are now only two commits - everything else was already in develop. No other changes, so there should not be any need to start from scratch.

nbougalis commented:
Also, a general comment: do `Transactor` derived classes still need a variadic template constructor? It seems that they only take an `ApplyContext` now.

ximinez commented:
@nbougalis 
> Also, a general comment: do Transactor derived classes still need a variadic template constructor? It seems that they only take an ApplyContext now.

**Need?** Strictly speaking, no, probably not. However, as I understand it, the intention of the variadic template constructors on the derived classes was to allow the derived classes to blindly pass all the parameters to the base `Transactor` without knowing or caring what they are, and, maybe more significantly, to allow those parameters to be changed without being forced to change the ctors for 10 derived classes every time we wanted to tweak the `Transactor` ctor. I would argue that this intention holds whether `Transactor` takes 4 parameters or 1... or 0.

I'd rather leave it variadic, unless there is a compelling reason not to.

vinniefalco commented:
@ximinez I think its safe to not use the variadic, the template argument was added before we had the `ApplyContext`.

ximinez commented:
There's enough smell coming from this that @vinniefalco and I are going to reexamine signature checking and preflight overall.

ximinez commented:
Ok, I think this is as good as it's gonna get. :)  The `HashRouter.sigVerify()` still has room for improvement, but these changes plug the hole that is the `mutable` field in `STTx`. We'll address a better, more general solution in a future change.

scottschurr commented:
Thanks for reducing the number of function objects in flight.  If it's as good as it gets then it's as good as it gets :-). :+1:

sublimator commented:
lol, love it: `[FOLD] Ok, fine, no more Transactor variadic ctors`

nbougalis commented:
:+1:

ximinez commented:
Squashed.
